File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
import sys

input = sys.stdin.readline
write = sys.stdout.write

def solve():
    n = int(input().rstrip())
    if not n: return
    
    wine = [0] * 10001
    for i in range(1, n + 1):
        wine[i] = int(input().rstrip())
        
    dp = [0] * 10001
    dp[1] = wine[1]
    
    if n > 1:
        dp[2] = wine[1] + wine[2]
        
    for i in range(3, n + 1):
        dp[i] = max(dp[i-1], dp[i-2] + wine[i], dp[i-3] + wine[i-1] + wine[i])
        
    write(str(dp[n]) + "\n")

if __name__ == "__main__":
    solve()

File: 백준/Silver/2156. 포도주 시식/포도주 시식.cc
#include <stdio.h>

#define MAX 10001

int main(void) {
	int n = 0;
	if (scanf("%d", &n) != 1) return 0;

	int wine[MAX] = { 0 };
	for (int i = 1; i <= n; i++) {
		if (scanf("%d", &wine[i]) != 1) break;
	}

	int dp[MAX] = { 0 };
	dp[1] = wine[1];

	if (n > 1) dp[2] = wine[1] + wine[2];

	for (int i = 3; i <= n; i++) {
		int max_val = dp[i - 1];
		if (max_val < dp[i - 2] + wine[i]) max_val = dp[i - 2] + wine[i];
		if (max_val < dp[i - 3] + wine[i - 1] + wine[i]) max_val = dp[i - 3] + wine[i - 1] + wine[i];

		dp[i] = max_val;
	}

	printf("%d\n", dp[n]);

	return 0;
}